code/modules/client/preferences.dm

```diff
@@ -808,6 +880,61 @@ GLOBAL_LIST_EMPTY(preferences_datums)
 				SetChoices(user)
 		return 1
 
+	else if(href_list["preference"] == "trait")
+		switch(href_list["task"])
+			if("close")
+				user << browse(null, "window=mob_occupation")
+				ShowChoices(user)
+			if("update")
+				var/trait = href_list["trait"]
+				var/value = SStraits.trait_points[trait]
+				if(value == 0)
+					if(trait in neutral_traits)
+						neutral_traits -= trait
+						all_traits -= trait
+					else
+						if(all_traits.len >= MAX_TRAITS)
+							to_chat(user, "<span class='warning'>You can't have more than [MAX_TRAITS] traits!</span>")
+							return
+						neutral_traits += trait
+						all_traits += trait
+				else
+					var/balance = GetTraitBalance()
+					if(trait in positive_traits)
+						positive_traits -= trait
+						all_traits -= trait
+					else if(trait in negative_traits)
+						if(balance + value < 0)

```

Comment:
The whole trait composition should be checked once instead of this giant IF tree per trait change in my opinion.

Reply:
What?

Reply:
Just do it once when all traits are to be saved instead of every change. Would make it easier to read.

Reply:
I'm not sure I follow... as in, cache the number and then update it when traits are saved, or don't change the number at all until they're saved?

Reply:
Just have a single validation function that checks the overall state.
```
DoWhatever() //Or after any other operation
Validate() //This will show some appropriate error messages. (Alerts are kinda clunky, better show it on the actual interface which parts are wrong)
..

if(Validate())
 Save()

```

Reply:
I'm having some issues getting this to work, by way of it not assigning positive traits at all. Here are the procs I'm using:

```
	else if(href_list["preference"] == "trait")
		if(SSticker.HasRoundStarted() && !isnewplayer(user))
			to_chat(user, "<span class='danger'>The round has already started. Please wait until next round to set up your traits!</span>")
			return
		switch(href_list["task"])
			if("close")
				user << browse(null, "window=mob_occupation")
				ShowChoices(user)
			if("update")
				var/trait = href_list["trait"]
				var/value = SStraits.trait_points[trait]
				if(ValidateTraits(trait, value))
					SaveTraits(trait, value)
				SetTraits(user)
```

ValidateTraits:
```
/datum/preferences/proc/ValidateTraits(trait, value) //ensures that you can take or lose a trait, returning FALSE if you can't
	if(isnull(value))
		return
	var/bal = GetTraitBalance()
	if(trait in all_traits && bal + value < 0)
		to_chat(usr, "<span class='warning'>Refunding this would cause your balance to go below zero!</span>")
		return
	else if(bal - value < 0)
		to_chat(usr, "<span class='warning'>You don't have the balance to take this!</span>")
		return
	else if(all_traits.len >= MAX_TRAITS)
		to_chat(usr, "<span class='warning'>You can only have [MAX_TRAITS] total traits!</span>")
		return
	return TRUE
```

SaveTraits:
```
/datum/preferences/proc/SaveTraits(trait, value) //adds the provided trait to the trait list
	var/list/category = list()
	if(value < 0)
		category = negative_traits
		to_chat(world, "Neg cat")
	else if(value == 0)
		category = neutral_traits
		to_chat(world, "Neu cat")
	else if(value > 0)
		category = positive_traits
		to_chat(world, "Pos cat")
	if(!category.len) //invalid value
		to_chat(world, "No cat")
		return
	if(trait in category)
		category -= trait
		all_traits -= trait
		to_chat(world, "Remove trait")
	else
		category += trait
		all_traits += trait
		to_chat(world, "Add trait")
```

The issue seems to be stemming in SaveTraits, as ValidateTraits passes fine. The `to_chat(world, X)` are for debug purposes. Can you see any immediate errors from just skimming through that might cause that?